### PR TITLE
disttask: refine dispatcher

### PR DESCRIPTION
**Pull request by ywqzzy**
### What problem does this PR solve?

Issue Number: ref #41495

Problem Summary:

Currently, the `dispatcher` use 2 for loop to dispatch and monitor tasks.

There is no abstraction for `dispatcher` and `dispatcherManager`.
1. `dispatcher` only handles one task.
2. `dispatcherManager` manages a bunch of `dispatcher`s and transfer submitted tasks to different `dispatcher`s.

### What is changed and how it works?
1.  remove `DetectTaskLoop`.
2. Don't call processFlow in `DispatchTaskLoop`, make it as light weight as possible.
3. add `monitorTask` which monitor the task status.
4. add `scheduleTask` which will call schedule one task with a for loop.
5. add `executeTask` interface which will trigger one goroutine to call scheduleTask.

Abstract `dispatcher` struct which only schedule and execute one task.

`dispatcher` will provide interface like:
1. `scheduleTask`: schedules the execution of subtasks.
2. `monitorTask`: check if the step is finished.
3. `GenerateDistPlan`: generate the task plan.

The `dispatcher`  do the following job step by step:
1. Generate distributed plan which have multiple steps.
2. dispatch first step's subtasks.
3. monitor first step's subtasks status and decide if going to next step. If next step, go to 2. If all steps subtasks finished, mark task finished.

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: domain/domain.go [hunk 1]
```diff
@@ -1707,12 +1707,12 @@ func (do *Domain) distTaskFrameworkLoop(ctx context.Context, taskManager *storag
...
 			return
 		}
-		newDispatch, err := dispatcher.NewDispatcher(ctx, taskManager)
+		newDispatcherPool, err := dispatcher.NewDispatcherPool(ctx, taskManager)
 		if err != nil {
 			logutil.BgLogger().Error("failed to create a disttask dispatcher", zap.Error(err))
 			return
 		}
-		dispatch = newDispatch
+		dispatch = newDispatcherPool
```
Comment (Benjamin2037) [inline domain/domain.go:1715]:
```suggestion
		dispatcher = newDispatcherPool
```

File: domain/domain.go [hunk 2]
```diff
@@ -1702,24 +1702,24 @@ func (do *Domain) distTaskFrameworkLoop(ctx context.Context, taskManager *storag
 		logutil.BgLogger().Info("dist task scheduler stopped")
 	}()
 
-	var dispatch dispatcher.Dispatch
+	var dispatcherPool *dispatcher.Pool
 	startDispatchIfNeeded := func() {
-		if dispatch != nil {
+		if dispatcherPool != nil {
```
Comment (Benjamin2037) [inline domain/domain.go:1707]: Since dispatcherPool is a point type, and here will check if dispatcherPool is nil. So you need init point value.

Comment (Benjamin2037) [inline domain/domain.go:1707]: Could we not use this kind of method to check whether the dispatcherPool is inited. I thought one struct with one IsInit bool type field is better.

**Review by Benjamin2037 — APPROVED** [review]
LGTM

File: disttask/framework/dispatcher/dispatcher_pool.go [hunk 3]
```diff
@@ -110,54 +111,56 @@ type dispatcher struct {
 	detectPendingGTaskCh chan *proto.Task
 }
 
-// NewDispatcher creates a dispatcher struct.
-func NewDispatcher(ctx context.Context, taskTable *storage.TaskManager) (Dispatch, error) {
-	dispatcher := &dispatcher{
+// NewPool creates a Pool struct.
```
Comment (zimulala) [inline disttask/framework/dispatcher/dispatcher_pool.go:114]: I think we needn't change this name to xxPool. 
Feel general xxPool, after the creation can have getXX, putXX and other interfaces. I don't think it's necessary here.

File: disttask/framework/dispatcher/dispatcher.go [hunk 4]
```diff
@@ -111,35 +115,43 @@ type dispatcher struct {
...
-// Stop implements Dispatch.Stop interface.
+// Stop implements dispatcher.Stop interface.
 func (d *dispatcher) Stop() {
 	d.cancel()
 	d.gPool.ReleaseAndWait()
 	d.wg.Wait()
+	d.inited = false
 }
 
+func (d *dispatcher) Inited() bool {
```
Comment (zimulala) [inline disttask/framework/dispatcher/dispatcher.go:148]: We don't use this function. Could we remove it?

Comment (ywqzzy) [reply disttask/framework/dispatcher/dispatcher.go:148]: I think `Inited` is better for checking of the dispatcher is started.

File: disttask/framework/dispatcher/dispatcher_manager.go [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+
+// Manager manage a bunch of dispatchers.
+// Dispatcher schedule and monitor tasks.
+// The scheduling task number is limited by size of gPool.
+type Manager struct {
+	ctx     context.Context
+	cancel  context.CancelFunc
+	taskMgr *storage.TaskManager
+	wg      tidbutil.WaitGroupWrapper
+	gPool   *spool.Pool
```
Comment (GMHDBJD) [inline disttask/framework/dispatcher/dispatcher_manager.go:78]: What does gpool do now?

Comment (ywqzzy) [reply disttask/framework/dispatcher/dispatcher_manager.go:78]: > What does gpool do now?

Fixed

File: disttask/framework/dispatcher/dispatcher.go [hunk 6]
```diff
@@ -51,377 +48,276 @@ var (
...
-				}
-				d.setRunningGTask(gTask)
-				cnt++
-			}
-		}
-	}
+// ExecuteTask start to schedule a task
+func (d *dispatcher) ExecuteTask() {
+	// Using the pool with block, so it wouldn't return an error.
+	_ = d.gPool.Run(func() {
```
Comment (GMHDBJD) [inline disttask/framework/dispatcher/dispatcher.go:98]: how about move to dispatcher manager?
```
func (dm *DispatcherManager) StartDispathcer(){
      dm.gPool.Run(func() {
             d := newDispatcher()
             d.ExecuteTask()
             d.delRunningTask()
      })
} 
```
Then we no need finished channel.

**Review by GMHDBJD — APPROVED** [review]
LGTM

**Pull request merged**